components/script/dom/htmlimageelement.rs

```diff
@@ -356,20 +390,253 @@ impl HTMLImageElement {
+                            if self.matches_environment(x.value().to_string()) {
+                                ()
+                            } else {
+                                continue;
+                            }
+                        }
+                        _ => ()
+                    }
+
+                    // Step 4.7
+                    match element.get_attribute(&ns!(), &local_name!("sizes")) {
+                        Some(x) => {
+                            source_set.sourceSize =
+                                parse_a_sizes_attribute(DOMString::from_string(x.value().to_string()));
+                            self.set_value_source_size_list(& mut source_set.sourceSize);
+                        }
+                        _ => ()
+                    }
+
+                    // Step 4.8
+                    match element.get_attribute(&ns!(), &local_name!("type")) {
+                        Some(x) => {
+                            match get_mime_type_str(&x.value()) {
+                                Some(_valid_mime) => (),
+                                None => continue // Unsupported mime type
+                            }
+                        }
+                        _ => continue // Unknown mime type
+                    }
+
+                    // Step 4.9
+                    Self::normalise_source_densities(&mut source_set);
+
+                    // Step 4.10
+                    *self.source_set.borrow_mut() = source_set;
+                    return ;
+                }
+            }
+        }
+    }
+
+    fn set_value_source_size_list(&self, source_size_list: &mut SourceSizeList) {
+        let document = document_from_node(self);
+        let device = document.device().unwrap();
+        let quirks_mode = document.quirks_mode();
+        let length = source_size_list.evaluate(&device, quirks_mode);
+        if !source_size_list.value.is_some() {
+            source_size_list.value = Some(Length::from_px(length.to_f32_px()));
+        }
+    }
+
+    /// https://html.spec.whatwg.org/multipage/#matches-the-environment
+    fn matches_environment(&self, media_query: String) -> bool {
+        let document = document_from_node(self);
+        let device = document.device().unwrap();
+        let quirks_mode = document.quirks_mode();
+        let document_url = &document.url();
+        let context = ParserContext::new(
+            Origin::Author,
+            document_url,
+            Some(CssRuleType::Style),
+            ParsingMode::all(),
+            quirks_mode,
+            None,
+        );
+        let mut parserInput = ParserInput::new(&media_query);
+        let mut parser = Parser::new(&mut parserInput);
+        let condition = MediaCondition::parse(&context, &mut parser);
+
+        let result = match condition {
+            Ok(x) => x.matches(&device, quirks_mode),
+            _ => false,
+        };
+
+        if *(&media_query.trim().is_empty()) || result {
```

Comment:
And once it does that the `is_empty` check is not necessary, not the owned `String`.

Reply:
Do you mean that MediaList checks for empty string condition too? Or that dereferencing is not needed ?

Reply:
@jdm Does this mean that we have to evaluate the MediaList to generate the final bool result?

Reply:
Yeah, an empty `MediaList` evaluates to true:

```
window.matchMedia("").matches
<- true
```